docs/architecture/adr-022-custom-panic-handling.md

```diff
@@ -0,0 +1,56 @@
+# ADR 022: Custom baseapp panic handling
+
+## Changelog
+
+- 2020 Apr 24: Initial Draft
+
+## Context
+
+Current implementation of BaseApp does not allow developers to write custom error handlers in
+[runTx()](https://github.com/cosmos/cosmos-sdk/blob/bad4ca75f58b182f600396ca350ad844c18fc80b/baseapp/baseapp.go#L538)
+method. We think that this method can be more flexible and can give SDK users more options for customizations without
+the need to rewrite whole BaseApp. Also there's one special case for `sdk.ErrorOutOfGas` error which feels like dirty
+hack in non-flexible environment.
+
+We propose middleware-solution, which could help developers implement following cases:
+* add external logging (let's say sending reports to external services like Sentry);
+* call panic for specific error cases;
+
+It will also make `OutOfGas` case and `default` case one of the middlewares.
+
+## Decision
+
+### Design
+
+Instead of hardcoding custom error handling into BaseApp we suggest using set of middlewares which can be customized
+externally and will allow developers use as many custom error handlers as they want.
+
+Implementation is already proposed [here](https://github.com/cosmos/cosmos-sdk/pull/6053).
```

Comment:
Great idea linking to the PR for reference. But I'd recommend actually stating the proposed APIs and logic here (in addition). It can be generalized and pseudo-code if you wish 👍